### ISSUE TITLE
ignore_update_merges for rebases

### ISSUE THREAD

**Issue by KnisterPeter** [palantir/policy-bot#208]
Could the `ignore_update_merges` option be extended to ignore rebases which do not change any diff while rebasing?
This would greatly help to keep approvals if a PR gets outdated by target branch updates.

**Comment by bluekeyes** [comment]
I don't think this is something we can easily support given the current architecture. `ignore_update_merges` uses a cheap heuristic that only works for merge commits created in the UI as a way to avoid expensive API calls to list and compare commit tress. To support rebases, I think we'd have to:

1. Track pull request state to record both the old head SHA and the new head SHA after a rebase and force push. Policy Bot is currently stateless and I'm not sure if the GitHub API has a reliable way to look up the before and after state of a PR that was force pushed.
2. Compare the trees of the new and old commits to determine if any files in the PR changed.

That said, it's possible that there's a clever and cheap way to detect safe rebases that I haven't thought of yet.

**Comment by devinburnette** [comment]
@bluekeyes I'm also interested in this feature as it would eliminate a lot of back and forth during review after either merging main into a feature branch or rebasing a PR that's already been approved. A lot has changed since this issue was created, and I think github event hooks have improved. We should be able to compare the `before` and `after` commit trees from the pull request synchronize event https://docs.github.com/en/webhooks/webhook-events-and-payloads#pull_request

So maybe we could implement logic that only invalidates a review if the tree SHAs aren't the same between the `before` and `after` commits. I feel like this would be a relatively cheap way to do this, but let me know if there's anything I'm missing or not thinking about.

**Comment by bluekeyes** [comment]
Unfortunately, I think anything based on webhook data is not an option. Policy Bot's current design requires that you can evaluate a pull request at any time and compute the correct result, so we can't implement techniques that only work with the data from a specific webhook delivery. Instead, we need to operate on data you can get at any time from the standard APIs.

The events API may include details about when a PR was updated via a force-push, but I've found this unreliable in the past and it may also provide incomplete or summarized data.

Particularly for rebases, I still think Policy Bot would have to track its own state somewhere outside of GitHub and I'm reluctant to add an external datastore as a requirement for running the bot.

**Comment by devinburnette** [comment]
@bluekeyes ok, in that case, could we instead add the commit SHA for each approval candidate to the Candidate struct, and then just look up the tree SHAs for both the head of the PR and the commit for the approval candidate when we filter out the invalidated ones?

**Comment by devinburnette** [comment]
nvm, i don't think that works in the case of rebases.. hmm
